pkg/services/sqlstore/searchstore/filters.go

```diff
@@ -85,6 +85,14 @@ func (f FolderFilter) Where() (string, []interface{}) {
 	return sqlIDin("dashboard.folder_id", f.IDs)
 }
 
+type FolderUIDFilter struct {
+	UIDs []string
+}
+
+func (f FolderUIDFilter) Where() (string, []interface{}) {
+	return sqlUIDin("folder_uid", f.UIDs)
+}
```

Comment:
I'm struggling with generating the correct SQL filter here. Currently this is making:


```sql
SELECT 
  dashboard.id, 
  dashboard.uid, 
  dashboard.title, 
  dashboard.slug, 
  dashboard_tag.term, 
  dashboard.is_folder, 
  dashboard.folder_id, 
  folder.uid AS folder_uid, 
  folder.slug AS folder_slug, 
  folder.title AS folder_title 
FROM 
  (
    SELECT 
      dashboard.id 
    FROM 
      dashboard 
    WHERE 
      (
        NOT dashboard.is_folder 
        OR dashboard.is_folder
      ) 
      AND dashboard.org_id = ? 
      AND folder_uid IN (?)  -- This is my additional folder UID filter
    ORDER BY 
      dashboard.title ASC 
    LIMIT 
      1000 OFFSET 0
  ) AS ids 
  INNER JOIN dashboard ON ids.id = dashboard.id 
  LEFT OUTER JOIN dashboard AS folder ON folder.id = dashboard.folder_id 
  LEFT OUTER JOIN dashboard_tag ON dashboard.id = dashboard_tag.dashboard_id 
ORDER BY 
  dashboard.title ASC
```

But after trying various things (`dashboard.folder_uid`, `folder_uid`, `folder.uid`), I think I've realised that the folder UID just isn't available in this subquery.

Reply:
I pushed a fix

<details>
<summary>Execution plan for example query searching by UID:</summary>

```
 mysql> EXPLAIN 
    -> SELECT
    ->                         dashboard.id,
    ->                         dashboard.uid,
    ->                         dashboard.title,
    ->                         dashboard.slug,
    ->                         dashboard_tag.term,
    ->                         dashboard.is_folder,
    ->                         dashboard.folder_id,
    ->                         folder.uid AS folder_uid,
    ->                         folder.slug AS folder_slug,
    ->                         folder.title AS folder_title  FROM ( SELECT dashboard.id FROM dashboard WHERE (NOT dashboard.is_folder OR dashboard.is_folder) AND dashboard.org_id=1 AND dashboard.folder_id IN (SELECT id FROM dashboard WHERE is_folder = 1 AND org_id = 1 AND uid IN ("bb138916-d6d1-4732-b15b-be462cd85ac0")) ORDER BY dashboard.title ASC LIMIT 1000 OFFSET 0) AS ids
    ->                 INNER JOIN dashboard ON ids.id = dashboard.id
    -> LEFT OUTER JOIN dashboard AS folder ON folder.id = dashboard.folder_id
    ->                 LEFT OUTER JOIN dashboard_tag ON dashboard.id = dashboard_tag.dashboard_id
    ->  ORDER BY dashboard.title ASC;
+----+-------------+---------------+------------+--------+---------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------+-----------------------------+------+----------+------------------------------------+
| id | select_type | table         | partitions | type   | possible_keys                                                                                                                                     | key                                  | key_len | ref                         | rows | filtered | Extra                              |
+----+-------------+---------------+------------+--------+---------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------+-----------------------------+------+----------+------------------------------------+
|  1 | PRIMARY     | <derived2>    | NULL       | ALL    | NULL                                                                                                                                              | NULL                                 | NULL    | NULL                        |  100 |   100.00 | Using temporary; Using filesort    |
|  1 | PRIMARY     | dashboard     | NULL       | eq_ref | PRIMARY                                                                                                                                           | PRIMARY                              | 8       | ids.id                      |    1 |   100.00 | NULL                               |
|  1 | PRIMARY     | folder        | NULL       | eq_ref | PRIMARY                                                                                                                                           | PRIMARY                              | 8       | grafana.dashboard.folder_id |    1 |   100.00 | NULL                               |
|  1 | PRIMARY     | dashboard_tag | NULL       | ref    | IDX_dashboard_tag_dashboard_id                                                                                                                    | IDX_dashboard_tag_dashboard_id       | 8       | ids.id                      |    2 |   100.00 | NULL                               |
|  2 | DERIVED     | dashboard     | NULL       | const  | PRIMARY,UQE_dashboard_org_id_folder_id_title,UQE_dashboard_org_id_uid,IDX_dashboard_org_id,IDX_dashboard_org_id_plugin_id,IDX_dashboard_is_folder | UQE_dashboard_org_id_uid             | 171     | const,const                 |    1 |   100.00 | Using filesort                     |
|  2 | DERIVED     | dashboard     | NULL       | ref    | UQE_dashboard_org_id_folder_id_title,UQE_dashboard_org_id_uid,IDX_dashboard_org_id,IDX_dashboard_org_id_plugin_id                                 | UQE_dashboard_org_id_folder_id_title | 16      | const,const                 |  110 |    91.00 | Using index condition; Using where |
+----+-------------+---------------+------------+--------+---------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------+-----------------------------+------+----------+------------------------------------+
6 rows in set, 1 warning (0.01 sec)
```
</details>

For comparison I provide the respective execution plan for the existing query (searching by ID)
<details>
<summary>Execution plan for example query searching by ID :</summary>

```
mysql> EXPLAIN SELECT
    ->                         dashboard.id,
    ->                         dashboard.uid,
    ->                         dashboard.title,
    ->                         dashboard.slug,
    ->                         dashboard_tag.term,
    ->                         dashboard.is_folder,
    ->                         dashboard.folder_id,
    ->                         folder.uid AS folder_uid,
    ->                         folder.slug AS folder_slug,
    ->                         folder.title AS folder_title  FROM ( SELECT dashboard.id FROM dashboard WHERE (NOT dashboard.is_folder OR dashboard.is_folder) AND dashboard.org_id=1 AND dashboard.folder_id IN (1) ORDER BY dashboard.title ASC LIMIT 1000 OFFSET 0) AS ids
    ->                 INNER JOIN dashboard ON ids.id = dashboard.id
    -> LEFT OUTER JOIN dashboard AS folder ON folder.id = dashboard.folder_id
    ->                 LEFT OUTER JOIN dashboard_tag ON dashboard.id = dashboard_tag.dashboard_id
    ->  ORDER BY dashboard.title ASC
    -> ;
+----+-------------+---------------+------------+--------+-------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------+-----------------------------+------+----------+------------------------------------+
| id | select_type | table         | partitions | type   | possible_keys                                                                                                     | key                                  | key_len | ref                         | rows | filtered | Extra                              |
+----+-------------+---------------+------------+--------+-------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------+-----------------------------+------+----------+------------------------------------+
|  1 | PRIMARY     | <derived2>    | NULL       | ALL    | NULL                                                                                                              | NULL                                 | NULL    | NULL                        |  100 |   100.00 | Using temporary; Using filesort    |
|  1 | PRIMARY     | dashboard     | NULL       | eq_ref | PRIMARY                                                                                                           | PRIMARY                              | 8       | ids.id                      |    1 |   100.00 | NULL                               |
|  1 | PRIMARY     | folder        | NULL       | eq_ref | PRIMARY                                                                                                           | PRIMARY                              | 8       | grafana.dashboard.folder_id |    1 |   100.00 | NULL                               |
|  1 | PRIMARY     | dashboard_tag | NULL       | ref    | IDX_dashboard_tag_dashboard_id                                                                                    | IDX_dashboard_tag_dashboard_id       | 8       | ids.id                      |    2 |   100.00 | NULL                               |
|  2 | DERIVED     | dashboard     | NULL       | ref    | UQE_dashboard_org_id_folder_id_title,UQE_dashboard_org_id_uid,IDX_dashboard_org_id,IDX_dashboard_org_id_plugin_id | UQE_dashboard_org_id_folder_id_title | 16      | const,const                 |  110 |    91.00 | Using index condition; Using where |
+----+-------------+---------------+------------+--------+-------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------+-----------------------------+------+----------+------------------------------------+
5 rows in set, 1 warning (0.02 sec)
```
</details>